### ISSUE TITLE
After isinstance check, type bounds are erased?

### ISSUE THREAD

**Issue by DylanLukes** [microsoft/pyright#8407]
This might admittedly be intended behavior, but it is very non-intuitive. It appears to me that at minimum the `isinstance` check _should_ propagate onwards to further checking that the types of `other.t1` and `other.t2` are bounded above by `Type`, in the same manner as occurs for `self.t1` and `self.t2`.

### Example 

```py
from typing import Any, Protocol, reveal_type


class TypeEquatable(Protocol):
    def __eq__(self, other: Any) -> bool:
        return type(self) == type(other)


class Type(TypeEquatable):
    pass

class Union[T1: Type, T2: Type](Type):
    def __init__(self, t1: T1, t2: T2) -> None:
        self.t1 = t1
        self.t2 = t2

    def __eq__(self, other: Any) -> bool:
        if isinstance(other, Union):
            reveal_type(self.t1)
            reveal_type(other.t1)

            return self.t1 == other.t1 and self.t2 == other.t2 \
                or self.t1 == other.t2 and self.t2 == other.t1

        return False
```

### Output

```sh
pyright --strict scratch.py
.../scratch.py
 .../scratch.py:28:25 - information: Type of "self.t1" is "T1@Union"
  .../scratch.py:29:25 - error: Type of "t1" is unknown (reportUnknownMemberType)
  .../scratch.py:29:25 - information: Type of "other.t1" is "Unknown"
  .../scratch.py:31:20 - error: Return type is unknown (reportUnknownVariableType)
  .../scratch.py:31:31 - error: Type of "t1" is unknown (reportUnknownMemberType)
  .../scratch.py:31:55 - error: Type of "t2" is unknown (reportUnknownMemberType)
  .../scratch.py:32:31 - error: Type of "t2" is unknown (reportUnknownMemberType)
  .../scratch.py:32:55 - error: Type of "t1" is unknown (reportUnknownMemberType)
  ```
  
  ### System Details
  
  This is with:
 - Python v3.12.3
 - Pyright  v1.1.371

And on macOS Sonoma 14.4.1

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

It appears that you have the `reportUnknownVariableType` and `reportUnknownMemberType` checks enabled. These are not enabled by default, so you have either chosen to enable them specifically or have opted in to "strict" type checking mode. When these are enabled, pyright will tell you when a type is partially unknown, which can lead to holes in type checking.

In your code example, `other` has the type `Any` prior to the `isinstance` type guard. After the type guard, the type of `other` is narrowed to `Union[Unknown, Unknown]` because there's no way to infer the proper type arguments in this case.

Your idea of substituting the upper bound type (`Type` in your code sample) is an interesting idea, but it would be feasible only in the case where the corresponding type parameter is covariant. In your code sample, `T1` and `T2` are invariant, which means that narrowing the type of `other` to `Union[Type, Type]` would frequently lead to other problems downstream.

The solution I recommend in this case is to use a user-defined type guard function that leverages the `TypeIs` special form. You can read more about this in [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeis).

**Comment by DylanLukes** [comment]
> In your code example, other has the type Any prior to the isinstance type guard. After the type guard, the type of other is narrowed to Union[Unknown, Unknown] because there's no way to infer the proper type arguments in this case.

This makes sense, thank you for explaining.

> The solution I recommend in this case is to use a user-defined type guard function that leverages the TypeIs special form. You can read more about this in [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeis).

I'll take a look at this. I was not aware it existed! :)
